xarray/plot/plot.py

```diff
@@ -419,10 +419,32 @@ def newplotfunc(darray, x=None, y=None, figsize=None, size=None,
         # Decide on a default for the colorbar before facetgrids
         if add_colorbar is None:
             add_colorbar = plotfunc.__name__ != 'contour'
+        imshow_rgb = plotfunc.__name__ == 'imshow' and \
+            darray.ndim == (3 + (row is not None) + (col is not None))
+        if imshow_rgb:
+            # Don't add a colorbar when showing an image with explicit colors
+            add_colorbar = False
+            # Manually stretch colors for robust cmap
+            if robust:
+                flat = darray.values.ravel(order='K')
+                flat = flat[~np.isnan(flat)]
+                if flat.size == 0:
+                    # All data will be masked, so skip percentile calculation
+                    vmin, vmax = 0, 1
+                if vmin is None:
+                    vmin = np.percentile(flat, ROBUST_PERCENTILE)
```

Comment:
I'm not quite sure that this handling of the robust argument makes sense. Does this resemble what vmin/vmax do by default on imshow?

Parts of this that seem strange to me:
- `robust=True` does more than merely clipping the data to the robust range, unlike the behavior with colormaps. The data gets extra normalization, even if it all already fell within the central 95%.
- all channels are treated together, rather than independently, including the alpha channel!

I'm sure this behavior is useful sometimes, but maybe it's better to leave it out for now. We could make robust=True an error for now, and potentially save color normalization for another keyword argument.

Reply:
This logic makes `robust=True` work exactly the same way on single-band and RGB greyscale images.  The handling is only distinct because the 'normalise to [0, 1]' step is implicit for single bands but must be explicit for RGB.

It does make sense to normalise RGB without considering or changing the alpha channel though; I'll play with this a little more.

As to usefulness, it's basically essential to sensibly visualise surface reflectance data (ie high-quality satellite images) - they're represented for analysis as albedo, ie fraction of light reflected, and if you display that directly it's a very dark and muddy image (example below; same scene as above).  So I'd strongly prefer to keep this in!

![download](https://user-images.githubusercontent.com/12229877/34325376-b60fb02e-e8e3-11e7-8dd4-1cb2d61598b6.png)

Reply:
This still feels strange to me. With colormaps, all `robust=True` does is explicitly set vmin and vmax based on the data. But vmin and vmax are entirely ignored here.

What if we interpreted vmin/vmax with rgb plots as setting the bounds for normalization? The defaults would be (0, 1) for floats and (smallest int, largest int) for int dtypes. Then, robust=True could again simply set vmin/vmax based on the data.

Reply:
I actually agree that the current behavior is strange and that your suggestion is an improvement (though I'd use [0, 255] for all ints) - but it's consistent with matplotlib!  My proposed plan has some extra steps though:

1. Merge this pull.  We will need the stretching machinery regardless, unless Xarray drops support for matplotlib<2.2
2. Patch matplotlib/matplotlib#9391 upstream.  (I'm already working on a patch for this)
3. Implement the new matplotlib interface in Xarray.  I *could* do that in this pull, but then if the matplotlib implementation doesn't match we'd have to break API compatibility - I'd really prefer to have a smaller API now and get it right the first time, later.

Reply:
> I could do that in this pull, but then if the matplotlib implementation doesn't match we'd have to break API compatibility - I'd really prefer to have a smaller API now and get it right the first time, later.

I agree. But I'm also concerned about the behavior for `robust=True` potentially needing to change to match upstream.

My preference would be to leave out support for `robust` with RGB plots until the matplotlib behavior lands. You could write a small helper function for RGB normalization (maybe even it include it in the docs as an example), e.g.,
```python
import xarray.ufuncs as xu

def robust_stretch(data):
  vmin, vmax = data.quantile(q=[0.02, 0.98])
  bounded = xu.minimum(xu.maximum(data, vmin), vmax)
  returned (bounded - vmin) / (vmax - vmin)
```
Then you could plot things like `arr.pipe(robust_stretch).plot.imshow(col='time', col_wrap=5)`.

It's a few more characters but not a terrible work around.